File: geekbang/algorithm/source-code/2/lc98 验证二叉搜索树.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root, long long l = -2147483648, long long r = 2147483647) {
        if (root == nullptr) return true;
        if (root->val < l || root->val > r) return false;
        return isValidBST(root->left, l, root->val - 1ll) && isValidBST(root->right, root->val + 1ll, r);
    }
};

File: geekbang/algorithm/source-code/2/lc98 验证二叉搜索树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.check(root, -2**31, 2**31-1)

    def check(self, root, rangeLeft, rangeRight):
        if not root:
            return True
        if root.val < rangeLeft or root.val > rangeRight:
            return False
        return self.check(root.left, rangeLeft, root.val - 1) and \
            self.check(root.right, root.val + 1, rangeRight)